### ISSUE TITLE
Melody sometimes removes classes from an element

### ISSUE THREAD

**Issue by iandevlin** [trivago/melody#87]
#### Explain the problem
Sometimes Melody removes an entire class listing from an element, causing styling issues.

#### Expected Behaviour
Melody shouldn't remove the `class` attribute from the element.

#### Actual Behaviour
Melody sometimes removes the `class` attribute

#### Steps to reproduce
It is difficult to reproduce as it doesn't happen consistently. I am not not sure under what conditions Melody does this.
But sometimes something like `<span class="class1 class2>Hello there</span>` is rendered as `<span>Hello there</span>`.
A reload will fix it, or using the classes filter: `<span class="{{ { base: 'class1 class2' } | classes }}">Hello there</span>`, but the `classes` filter shouldn't need to be used here as there are no conditional classes.

#### Provide a repository that reproduces issue if possible
N/A

#### Provide your Environment details
- Node version: v8.9.4

- Operating System: MacOS Mojave v10.14.1 

- Melody version: v0.5.1

**Comment by malte-wessel** [comment]
I remember a similar bug where a switch between static and dynamic properties lead to the mentioned situation:

```twig
<div>
    {% if foo %}
        <div class="foo"></div>
    {% else %}
        <div class="foo {{ bar ? 'bar' }}"></div>
    {% endif %}
</div>

```

**Comment by pago** [comment]
## Investigation

The example provided by @malte-wessel will generate the following code:

```
const _statics = [\\"class\\", \\"foo\\"];

_template.render = function (_context) {
    elementOpen(\\"div\\", null, null);

    if (_context.foo) {
        elementVoid(\\"div\\", \\"7*U2;JR\\", _statics);
    } else {
        elementVoid(\\"div\\", null, null, \\"class\\", \\"foo \\" + (_context.bar ? \\"bar\\" : \\"\\"));
    }

    elementClose(\\"div\\");
};
```

So the `div` with the static class will have a `key` while the second one with the dynamic `class` attribute will not.
First question for me would be: Why?

Continuing the search:
https://github.com/trivago/melody/blob/master/packages/melody-idom/src/virtual_elements.ts#L68
`elementOpen` only forwards the element name and the `key` to core `elementOpen` feature which can therefore not use the information that `static` attributes have gone missing to understand that we are not dealing with the same element anymore.

This leads to the very unfortunate situation which we can observe here:
https://github.com/trivago/melody/blob/master/packages/melody-idom/src/core.ts#L226
and which is even documented there:

```
// exisiting DOM element does not have a key
// which means we can hook onto it freely
```

That ultimately leads to the element being reused even though we're dealing with different elements. This would be triggered if we first had `foo == false` and then switched to `foo == true`.

To this point, we're not actually looking at a major issue since reusing the `div` might be a valuable thing to do. It might share more of its internals, allowing us to potentially reduce the amount of element creation/deletion needed.
The problem is, though, that we're not updating the attributes:
https://github.com/trivago/melody/blob/master/packages/melody-idom/src/virtual_elements.ts#L83

And sure enough the `staticsApplied` is true since we have touched the element before and we did add all static attributes (there were none). The *dynamic*  class attribute is no longer specified so it is removed and we end up with a div that doesn't have its class anymore.

## Identified problems
- Second element is generated without a key, thus Melody doesn't know its different
- Attempt to reuse fails because static attributes are not verified after the first time

## Suggested solution
Store and check the static attributes. If they have changed (by reference) we should invalidate them and re-apply them completely.

**Comment by ayusharma** [comment]
Fixed in v1.2.0
